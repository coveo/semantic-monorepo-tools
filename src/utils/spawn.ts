import { Debugger } from "debug";
import { spawn, SpawnOptions } from "node:child_process";
import defaultLogger from "./logger.js";

interface Options extends SpawnOptions {
  encoding?: "ascii" | "utf8" | "utf-8";
}

type spawnOutputs = {
  stdout: string;
  stderr: string;
};

export default function (
  command: string,
  args: string[],
  logger: Debugger = defaultLogger,
  options: Options = {},
): Promise<{ stdout: string; stderr: string }> {
  const loggableArgs = args.join(" ");
  logger?.('executing "%s %s"', command, loggableArgs);
  let stdout = "";
  let stderr = "";
  const { encoding, ...spawnOptions } = { encoding: "utf-8", ...options };

  return new Promise<spawnOutputs>((resolve, reject) => {
    const childProcess = spawn(command, args, spawnOptions);

    childProcess.stdout.on("data", (data) => {
      logger?.(`STDOUT: ${data}`);
      stdout += data.toString(encoding);
    });

    childProcess.stderr.on("data", (data) => {
      logger?.(`STDERR: ${data}`);
      stderr += data.toString(encoding);
    });

    childProcess.on("close", (code, signal) => {
      logger?.(
        `${command} ${loggableArgs} finished with code ${code} and signal '${signal}'`,
      );
      if (code) {
        reject({ stdout, stderr });
      } else {
        resolve({ stdout, stderr });
      }
    });
  });
}
