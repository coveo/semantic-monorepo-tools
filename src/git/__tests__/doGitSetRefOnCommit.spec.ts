import spawn from "../../utils/spawn.js";
import gitSetRef from "../doGitSetRefOnCommit.js";

jest.mock("../../utils/spawn.js");
jest.mock("../utils/gitLogger.js");

const mockedSpawn = jest.mocked(spawn);

describe("doGitSetRef()", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    mockedSpawn.mockResolvedValue({ stdout: "", stderr: "" });
  });

  describe.each(["--upload-pack", "--whatever --upload-pack"])(
    "with invalid params",
    (remote) => {
      it("should throw when called with a bad remote", () => {
        expect(
          gitSetRef(remote, "a", "b"),
        ).rejects.toThrowErrorMatchingSnapshot();
      });
      it("should throw when called with a bad commitSHA1", () => {
        expect(
          gitSetRef("a", "a", remote),
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    },
  );

  it("should append --force when force=true", async () => {
    await gitSetRef("origin", "a", "b", true);
    expect(mockedSpawn.mock.lastCall[1].pop()).toBe("--force");
  });

  it("should not append --force when force is unset", async () => {
    await gitSetRef("origin", "a", "b");
    expect(mockedSpawn.mock.lastCall[1].pop()).not.toBe("--force");
  });

  it("should not append --force when force=false", async () => {
    await gitSetRef("origin", "a", "b", false);
    expect(mockedSpawn.mock.lastCall[1].pop()).not.toBe("--force");
  });
});
