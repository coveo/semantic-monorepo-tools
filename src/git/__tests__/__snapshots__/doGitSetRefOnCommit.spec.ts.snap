// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`doGitSetRef() with invalid params should throw when called with a bad commitSHA1 1`] = `"invalid param:a,a,--upload-pack"`;

exports[`doGitSetRef() with invalid params should throw when called with a bad commitSHA1 2`] = `"invalid param:a,a,--whatever --upload-pack"`;

exports[`doGitSetRef() with invalid params should throw when called with a bad remote 1`] = `"invalid param:--upload-pack,a,b"`;

exports[`doGitSetRef() with invalid params should throw when called with a bad remote 2`] = `"invalid param:--whatever --upload-pack,a,b"`;
